python/lsst/daf/butler/_butler.py

```diff
@@ -1597,6 +1599,161 @@ def retrieveArtifacts(
+            `DatasetRef` within a Collection. When `None`, a `DatasetRef`
+            should be provided as the first argument.
+        full_check : `bool`, optional
+            If `True`, an additional check will be made for dataset artifact
+            existence. This will involve additional overhead due to the need
+            to query an external system. If `False` registry and datastore
+            will solely be asked if they know about the dataset but no
+            check for the artifact will be performed.
+        collections : Any, optional
+            Collections to be searched, overriding ``self.collections``.
+            Can be any of the types supported by the ``collections`` argument
+            to butler construction.
+        **kwargs
+            Additional keyword arguments used to augment or construct a
+            `DataCoordinate`.  See `DataCoordinate.standardize`
+            parameters.
+
+        Returns
+        -------
+        existence : `DatasetExistence`
+            Object indicating whether the dataset is known to registry and
+            datastore. Evaluates to `True` if the dataset is present and known
+            to both.
+        """
+        existence = DatasetExistence.UNRECOGNIZED
+
+        if isinstance(dataset_ref_or_type, DatasetRef):
+            ref = dataset_ref_or_type
+            registry_ref = self.registry.getDataset(dataset_ref_or_type.id)
+            if registry_ref is not None:
+                existence |= DatasetExistence.RECORDED
+
+                if dataset_ref_or_type != registry_ref:
+                    # This could mean that storage classes differ, so we should
+                    # check for that but use the registry ref for the rest of
+                    # the method.
+                    if registry_ref.is_compatible_with(dataset_ref_or_type):
+                        # Use the registry version from now on.
+                        ref = registry_ref
+                    else:
+                        raise ValueError(
+                            f"The ref given to exists() ({ref}) has the same dataset ID as one "
+                            f"in registry but has different incompatible values ({registry_ref})."
+                        )
+        else:
+            try:
+                ref = self._findDatasetRef(dataset_ref_or_type, data_id, collections=collections, **kwargs)
+            except (LookupError, TypeError, NoDefaultCollectionError):
+                return existence
+            existence |= DatasetExistence.RECORDED
+
+        if self.datastore.knows(ref):
+            existence |= DatasetExistence.DATASTORE
+
+        if full_check:
+            if self.datastore.exists(ref):
+                existence |= DatasetExistence._ARTIFACT
+        elif existence != DatasetExistence.UNRECOGNIZED:
+            # Do not add this flag if we have no other idea about a dataset.
+            existence |= DatasetExistence._ASSUMED
```

Comment:
Should this check be for `existence | DatasetExistence.DATASTORE`, does it make sense to have `_ASSUMED` without `DATASTORE`? Maybe whole `if full_check:` wants to be nested in `if self.datastore.knows(ref):`?

Reply:
It is possible for the file to exist but for datastore not to know about it. That's the reason we have to two independent states. This is how execution butler runs and how `trustGetRequest` works. `full_check` therefore can't be nested inside `knows`. In a "proper" datastore the `datastore.exists` will check if it knows first and return False regardless.

Reply:
I may have misunderstood what happens in trust mode.  Does that never return DATASET_KNOWS or ARTIFACT_ASSUMED?

I guess that makes sense but I worry that it exposes an implementation detail (whether the datastore is in trust mode).

Of course, that may not matter if we can get to the point where trust mode is only ever paired with QBB, not a full Butler, after execution butler is gone.  I'm hoping we can.

Reply:
`_ASSUMED` is a private flag entirely there to make things obvious internally. It's not a public state. In execution butler it will return False because it really doesn't know about the ref and full check will indicate that the artifact does exist. You will see extensive tests exploring all the different states. If `datastore.trustGetRequest` is something that `Butler.exists` is aware of then we could make it look on the file system regardless of `full_check` and add `DATASTORE` flag if the artifact exists.